### PR TITLE
[API] Add convenience methods for reading env and setting secrets from files

### PR DESCRIPTION
**Pull request by yaronha**
convenience methods for reading env and setting secrets from files
the env file should have lines in the form KEY=VALUE, comment line start with "#"

* `set_env_from_file()` for reading `.env` files, setting the OS environment and reloading MLRun config
* `project.set_secrets()` will read dict or secrets env file and store in the project secrets 
  * (note that MLRUN_DBPATH and V3IO_xxx vars will not be written to the project secrets)
* `function.set_envs()` now have a new param `file_path` to set the env from a `.env` file

example env file:
```
# this is an env file
V3IO_USERNAME=admin
V3IO_API=https://webapi.default-tenant.app.xxx.iguazio-cd1.com
V3IO_ACCESS_KEY=MYKEY123
MLRUN_DBPATH=https://mlrun-api.default-tenant.app.xxx.iguazio-cd1.com
AWS_ACCESS_KEY_ID=XXXX
AWS_SECRET_ACCESS_KEY=YYYY
```

usage:
```python
# set the env vars from a file and also return the results as a dict (e.g. for using in a function)
env_dict = mlrun.set_env_from_file(env_path, return_dict=True)

# read env vars from dict or file and set as project secrets (plus set the local env)
project.set_secrets({"SECRET1": "value"})
project.set_secrets(file_path=env_file)

# copy env from file into a function spec
function.set_envs(file_path=env_file)
```

### CODE REVIEW

File: mlrun/__init__.py [hunk 1]
```diff
@@ -160,3 +160,81 @@ def get_sample_path(subpath=""):
...
+    :param to_dict:  set to True to return the env as a dict
+    :param set_env:  set to False to skip updating the current OS env
+    :return: None or env dict
+    """
+    env_vars = {}
+    body = get_dataitem(env_file).get(encoding="utf-8")
+    for line in body.splitlines():
+        if line.startswith("#") or not line.strip():
+            continue
+        key, value = line.strip().split("=", 1)
```
Comment (theSaarco) [inline mlrun/__init__.py:194]: Should check that `key` has any value, otherwise lines like `=something` will end up with an empty environment variable. It's perfectly valid for `value` to be empty, but I don't see any sense in having an empty key.

File: mlrun/__init__.py [hunk 2]
```diff
@@ -160,3 +160,81 @@ def get_sample_path(subpath=""):
...
+            environ[key] = value  # Load to local environ
+            if key == "MLRUN_DBPATH":
+                mlconf.dbpath = value
+            if key == "V3IO_API":
+                mlconf.v3io_api = value
+        env_vars[key] = value
+    return env_vars if to_dict else None
+
+
+def file_to_project_secrets(env_file, project=None, set_env=False, provider=None):
```
Comment (theSaarco) [inline mlrun/__init__.py:205]: Why have a default for `project`? Having `project=None` will explode when looking for `project_name`, so I don't see the value in having this default here - user should always call with a project.
Ideally, also add valid types for project: `project: Union[str, MLRunProject, MLRunProjectLegacy]` or something like that.

File: tests/assets/envfile [hunk 3]
```diff
@@ -0,0 +1,4 @@
+ENV_ARG1=123
+
+# comment
+ENV_ARG2=abc
```
Comment (theSaarco) [inline tests/assets/envfile:4]: Add some more validations here - 
```
=something
ENV=
=
```

File: mlrun/__init__.py [hunk 4]
```diff
@@ -160,3 +160,81 @@ def get_sample_path(subpath=""):
...
+    :param set_env:   set to True to also configure the local OS env
+    :param provider:  MLRun secrets provider
+    """
+    env_vars = env_from_file(env_file, to_dict=True, set_env=set_env)
+    # drop V3IO paths/credentials and MLrun service API address
+    env_vars = {
+        key: val
+        for key, val in env_vars.items()
+        if key != "MLRUN_DBPATH" and not key.startswith("V3IO_")
+    }
```
Comment (theSaarco) [inline mlrun/__init__.py:236]: I think in this case it also makes sense to filter out secrets whose value is empty (in the `env_from_file` we should just filter empty keys) - there's no value in adding a secret with just a key.

File: tests/system/api/assets/envfile [hunk 5]
```diff
@@ -0,0 +1,6 @@
+ENV_ARG1=123
+
+# comment
+ENV_ARG2=abc
+# verify V3IO vars are not stored
+V3IO_USERNAME=xyz
```
Comment (theSaarco) [inline tests/system/api/assets/envfile:6]: Same comment - add more validations.

File: tests/system/api/test_secrets.py [hunk 6]
```diff
@@ -187,3 +187,15 @@ def test_k8s_project_secrets_with_runtime(self):
 
         # Cleanup secrets
         self._run_db.delete_project_secrets(self.project_name, provider="kubernetes")
+
+    def test_file_to_project_secrets(self):
+        env_file = str(pathlib.Path(__file__).parent / "assets" / "envfile")
+        self._run_db.delete_project_secrets(self.project_name, provider="kubernetes")
+        mlrun.file_to_project_secrets(env_file, self.project_name)
```
Comment (theSaarco) [inline tests/system/api/test_secrets.py:194]: Better call with `set_env=False` to make sure we're not polluting the env for other tests.

File: tests/test_convenince_methods.py [hunk 7]
```diff
@@ -1,11 +1,24 @@
...
     with pytest.raises(mlrun.errors.MLRunInvalidArgumentError):
         mlrun.set_environment(project=invalid_name)
+
+
+def test_env_from_file():
+    env_path = str(assets_path / "envfile")
+    env_dict = mlrun.env_from_file(env_path, to_dict=True)
+    assert env_dict == {"ENV_ARG1": "123", "ENV_ARG2": "abc"}
+    for key, value in env_dict.items():
+        assert os.environ[key] == value
```
Comment (theSaarco) [inline tests/test_convenince_methods.py:24]: Should cleanup the env, otherwise even running this same test twice doesn't really check anything here.

**Review by theSaarco — CHANGES_REQUESTED** [review]
Looks good - really handy functionality. Have a few small things that need fixing.

File: tests/assets/envfile (hunk 3 above)
Comment (yaronha) [reply tests/assets/envfile:4]: ok, `ENV=` is valid IMO

File: mlrun/__init__.py (hunk 4 above)
Comment (yaronha) [reply mlrun/__init__.py:236]: empty key is valid IMO, if u dont want them dont have lines w key and "" value

File: mlrun/__init__.py (hunk 1 above)
Comment (yaronha) [reply mlrun/__init__.py:194]: ok

File: mlrun/__init__.py [hunk 8]
```diff
@@ -160,3 +160,81 @@ def get_sample_path(subpath=""):
     if subpath:
         samples_path = path.join(samples_path, subpath.lstrip("/"))
     return samples_path
+
+
+def env_from_file(env_file, to_dict=False, set_env=True):
```
Comment (Hedingber) [inline mlrun/__init__.py:165]: Function name usually start with a verb -> `set_env_from_file`

File: mlrun/__init__.py [hunk 9]
```diff
@@ -160,3 +160,81 @@ def get_sample_path(subpath=""):
...
+        AWS_ACCESS_KEY_ID-XXXX
+        AWS_SECRET_ACCESS_KEY=YYYY
+
+    usage::
+
+        # set the env vars from a file + return the results as a dict
+        env_dict = mlrun.env_from_file(env_path, to_dict=True)
+
+    :param env_file: path/url to env file
+    :param to_dict:  set to True to return the env as a dict
```
Comment (Hedingber) [inline mlrun/__init__.py:185]: Why is this needed ? simply always return it

File: mlrun/__init__.py [hunk 10]
```diff
@@ -160,3 +160,81 @@ def get_sample_path(subpath=""):
...
+    :param set_env:  set to False to skip updating the current OS env
```
Comment (Hedingber) [inline mlrun/__init__.py:186]: This param also seems redundant to me, keep things simple, what sense there is in doing something like `set_env_from_file(set_env=False)` if you need it internally, have an internal function doing that for you, no need to expose it outside.
Also, IMO this function should just become a classmethod of the mlrun config - `mlrun.mlconf.from_env_file(file_path)` or `mlrun.mlconf.load_env_file(file_path)` or something like that

**Review by Hedingber — CHANGES_REQUESTED** [review]

File: mlrun/__init__.py (hunk 10 above)
Comment (yaronha) [reply mlrun/__init__.py:186]: assume i want to copy env file to set local env and/or set the function env (`fn.set_envs()`), i rather do it in one call

Comment (yaronha) [reply mlrun/__init__.py:186]: also i think its an essential func (why have it under `mlrun.` ) and its not specific to conf (can use it to load S3 env, user env..)

File: mlrun/__init__.py (hunk 9 above)
Comment (yaronha) [reply mlrun/__init__.py:185]: in many cases u use it in the Notebook, u dont want the creds to be printed to the output cell

File: mlrun/__init__.py (hunk 8 above)
Comment (yaronha) [reply mlrun/__init__.py:165]: ok

File: mlrun/__init__.py (hunk 2 above)
Comment (yaronha) [reply mlrun/__init__.py:205]: ok

File: tests/test_convenince_methods.py (hunk 7 above)
Comment (yaronha) [reply tests/test_convenince_methods.py:24]: ok

File: tests/system/api/assets/envfile (hunk 5 above)
Comment (yaronha) [reply tests/system/api/assets/envfile:6]: added to the underline func

**Review by Hedingber — APPROVED** [review]
